weston-ivi-shell/src/ivi-controller.c

```diff
@@ -476,101 +437,82 @@ ivi_weston_compositor_read_presentation_clock(
-    // assuming ABGR32 is always written by surface_dump
-    uint32_t format = WL_SHM_FORMAT_ABGR8888;
     struct wl_resource *screenshot;
     struct timespec stamp;
     uint32_t stamp_ms;
-    int fd;
-
-    screenshot =
-        wl_resource_create(client, &ivi_screenshot_interface, 1, screenshot_id);
+    void *shm_buff_data = NULL;
+    struct weston_buffer *weston_buffer = NULL;
 
+    screenshot = wl_resource_create(client,
+            &ivi_screenshot_interface, 2, screenshot_id);
     if (screenshot == NULL) {
         wl_client_post_no_memory(client);
         return;
     }
 
     layout_surface = lyt->get_surface_from_id(surface_id);
     if (!layout_surface) {
-        ivi_screenshot_send_error(
-            screenshot, IVI_SCREENSHOT_ERROR_NO_SURFACE,
-            "surface_screenshot: the surface with given id does not exist");
+        ivi_screenshot_send_error(screenshot, IVI_SCREENSHOT_ERROR_NO_SURFACE, 
+                "surface_screenshot: the surface with given id does not exist");
         goto err;
     }
 
-    result = lyt->surface_get_size(layout_surface, &width,
-                                   &height, &stride);
+    result = lyt->surface_get_size(layout_surface, &width, &height, &stride);
     if ((result != IVI_SUCCEEDED) || !width || !height || !stride) {
-        ivi_screenshot_send_error(
-            screenshot, IVI_SCREENSHOT_ERROR_NO_CONTENT,
-            "surface_screenshot: surface does not have content");
+        ivi_screenshot_send_error(screenshot, IVI_SCREENSHOT_ERROR_NO_CONTENT, 
+                "surface_screenshot: surface does not have content");
         goto err;
     }
 
-    size = stride * height;
-
-    fd = create_screenshot_file(size);
-    if (fd < 0) {
-        weston_log(
-            "surface_screenshot: failed to create file of %d bytes: %m\n",
-            size);
-        ivi_screenshot_send_error(
-            screenshot, IVI_SCREENSHOT_ERROR_IO_ERROR,
-            "failed to create screenshot file");
+    /* get the weston buffer from buffer resource, then verify */
+    weston_buffer = weston_buffer_from_resource(compositor, buffer_resource);
+    if ( (weston_buffer == NULL) ||
+            (weston_buffer->type != WESTON_BUFFER_SHM) ||
+            /* assuming ABGR32 is always written by surface_dump.
+             * ABGR32 may not support by rederer to create a shm buffer.
+             * So, just check the bytes per pixel must be 4 here.
+             */
+            ((wl_shm_buffer_get_stride(weston_buffer->shm_buffer) /
+                (wl_shm_buffer_get_width(weston_buffer->shm_buffer))) != 4) ||
+            (wl_shm_buffer_get_width(weston_buffer->shm_buffer) < width) ||
+            (wl_shm_buffer_get_height(weston_buffer->shm_buffer) < height)) {
+        ivi_screenshot_send_error(screenshot, IVI_SCREENSHOT_ERROR_BAD_BUFFER,
+                "bad buffer input");
         goto err;
     }
+    shm_buff_data = wl_shm_buffer_get_data(weston_buffer->shm_buffer);
```

Comment:
shm buffer access should be bounded by begin_access and end_access calls. Refer to gl-renderer code parts where shm_buffer access is carried out.

Reply:
Hi Harsha, 

I added the bounding for `surface_dump`, but get the assertion from weston:
`weston: ../src/wayland-shm.c:652: wl_shm_buffer_begin_access: Assertion 'sigbus_data->current_pool == NULL || sigbus_data->current_pool == pool' failed`.

I had took a look, then I can see:
- In surface_dump, it also calls to `wl_shm_buffer_begin_access` and `wl_shm_buffer_end_access` with input is `surface buffer`.
![weston_assertion](https://github.com/COVESA/wayland-ivi-extension/assets/113493212/b0eae15f-1f01-490d-873d-236a9a6a4df9)
- Look at the description of `wl_shm_buffer_end_access` on [here](https://gitlab.freedesktop.org/wayland/wayland/-/blob/1.20.0/src/wayland-shm.c#L612), we see, it only safe to nest calls, if all of them is calling a same buffer. 
- In the description, also mention the usage of this function, it seems, we should call it when start to read the buffer.

=> I think we don't need those APIs in our writing case. Let me know your opinion. thanks

Reply:
@khangtb1 ,

For me it looks like, even "ivi layout" surface_dump call could have used wl_buffer as the interface. Looks like for screenshot to a surface, adaptation to wl_buffer is missing in weston. Right since we have a write case, may be we can ignore the begin and end buffer. Never the less, I see the right approach is to have wl_buffer as argument for surface_dump function.